### PR TITLE
Testing manifesto

### PR DESCRIPTION
**Pull request by jjanvier**
The goal is this manifesto is to:
- help newcomers to understand how we test
- help current devs to know how we should test
- be the central source of truth regarding the team's choices about tests

This is a first version. Of course, it can/will/should evolve.

### CODE REVIEW

File: TESTING.md [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+
+
+## FAQ
+
+> I don't know if I should write a unit, an acceptance, an integration or an end-to-end test. What should I do?
+
+You should refer to the [ports and adapters architecture](#ports-and-adapters-architecture). Ask yourself, where your class is standing regarding the different layers. Then, you can refer to [this section](#the-relation-with-the-tests).
+
+> I'm afraid to write less end to end tests that before. Are you sure it's a good idea?
+
```
Comment (docteurklein) [inline TESTING.md]: even if he uses a different naming, this article is cool: http://blog.thecodewhisperer.com/permalink/integrated-tests-are-a-scam
also https://testing.googleblog.com/2015/04/just-say-no-to-more-end-to-end-tests.html

File: TESTING.md [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+> What is a service that performs I/O calls?
+
+Any service that uses an external system (relatively to your code). Can be considered as external systems: the file system, the system time, any system called via the network, a database or a search engine for instance. That means a Doctrine repository, which communicates with the database, is a service performing I/O calls.
+
+## Resources
+
+> How can I write useful and powerful Gherkin?
+
+[Modelling by Example Workshop](https://fr.slideshare.net/CiaranMcNulty/modelling-by-example-workshop-phpnw-2016) by Ciaran McNulty
```
Comment (docteurklein) [inline TESTING.md]: also http://stakeholderwhisperer.com/posts/2014/10/introducing-modelling-by-example

File: TESTING.md [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+- frontend: TODO
+
+### Acceptance - test a business use case or ensure a business rule
+
+Characteristics:
+
+- it tests several classes at the same time
+- it uses the business language, which means _Gherkin_ must be used
+- it describes a business use case or it ensures a business rule (it's not about UI, CLI or UX, neither about a text we should see)
+- it mocks only what we own
```
Comment (aRn0D) [inline TESTING.md]: We don't mock, we create fake objects (example InMemoryRepository) or we stub what we don't own (example HTTP call)

File: TESTING.md [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+- it mocks services performing [I/O calls](#faq)
```
Comment (aRn0D) [inline TESTING.md]: what is 'services performing'?

Comment (juliensnz) [reply TESTING.md]: @aRn0D "service performing I/O calls "

File: TESTING.md [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+- it tests only one class at a time
+- it performs no [I/O call](#faq)
+- it doesn't require any setup (no fixtures for instance)
+- it mocks only what we own
+- it "lives" in memory only
+
+Tools:
+
+- backend: PhpSpec
+- frontend: TODO
```
Comment (juliensnz) [inline TESTING.md]: jest+enzyme

File: TESTING.md [hunk 6]
```diff
@@ -0,0 +1,153 @@
...
+- backend: TODO
+- frontend: TODO
+
+### Acceptance - test a business use case or ensure a business rule
+
+Characteristics:
+
+- it tests several classes at the same time
+- it uses the business language, which means _Gherkin_ must be used
+- it describes a business use case or it ensures a business rule (it's not about UI, CLI or UX, neither about a text we should see)
```
Comment (juliensnz) [inline TESTING.md]: Could be interesting to discuss about this rule on frontend part. I have some examples to show you.

File: TESTING.md [hunk 7]
```diff
@@ -0,0 +1,153 @@
...
+- it mocks only what we own
+- it mocks services performing [I/O calls](#faq)
+- it "lives" in memory only
+
+Tools:
+
+- backend: Gherkin through Behat (no Mink, no Selenium)
+- frontend: Gherkin through TODO
```
Comment (juliensnz) [inline TESTING.md]: Gherkin through cucumberjs and puppeteer

File: TESTING.md [hunk 8]
```diff
@@ -0,0 +1,153 @@
...
+Characteristics:
+
+- it has no mock (it tests the real classes)
+- it may test several classes at the same time
+- it tests only services that perform [I/O calls](#faq)
+
+Tools:
+
+- backend: PhpUnit
+- frontend: TODO
```
Comment (juliensnz) [inline TESTING.md]: jest with puppeteer

File: TESTING.md [hunk 9]
```diff
@@ -0,0 +1,153 @@
...
+
+Tools:
+
+- backend: PhpSpec
+- frontend: TODO
+
+Examples:
+
+- backend: TODO
+- frontend: TODO
```
Comment (juliensnz) [inline TESTING.md]: "Test the dropdown component"

File: TESTING.md [hunk 10]
```diff
@@ -0,0 +1,153 @@
...
+### The ideal pyramid
+
+
+## FAQ
+
+> I don't know if I should write a unit, an acceptance, an integration or an end-to-end test. What should I do?
+
+You should refer to the [ports and adapters architecture](#ports-and-adapters-architecture). Ask yourself, where your class is standing regarding the different layers. Then, you can refer to [this section](#the-relation-with-the-tests).
+
+> I'm afraid to write less end to end tests that before. Are you sure it's a good idea?
```
Comment (juliensnz) [inline TESTING.md]: "than before"

File: TESTING.md [hunk 11]
```diff
@@ -0,0 +1,153 @@
...
+- frontend: test a fetcher performing HTTP calls
+
+### End to end - test the whole application
+
+Characteristics:
+
+- it tests the application as a whole (the backend and the frontend are tested at the same time)
+- it has no mock (it tests the real application)
+- it can require a complex setup (like a browser and Selenium for instance)
+- it tests nominal use cases
```
[the remark this answers is not in the document]
Comment (jjanvier) [reply TESTING.md]: fixed

File: TESTING.md (hunk 2 above)
Comment (jjanvier) [reply TESTING.md]: fixed

File: TESTING.md (hunk 10 above)
Comment (jjanvier) [reply TESTING.md]: fixed

File: TESTING.md (hunk 5 above)
Comment (jjanvier) [reply TESTING.md]: fixed

File: TESTING.md (hunk 7 above)
Comment (jjanvier) [reply TESTING.md]: fixed

File: TESTING.md (hunk 8 above)
Comment (jjanvier) [reply TESTING.md]: fixed

File: TESTING.md (hunk 4 above)
Comment (aRn0D) [reply TESTING.md]: So, it is not a mock, it is a stub!

File: TESTING.md [hunk 12]
```diff
@@ -0,0 +1,167 @@
...
+
+> I want to know more about that ports and adapters thing!
+
+[Hexagonal architecture](http://alistair.cockburn.us/Hexagonal%20architecture), original article by Alistair Cockburn
+
+[Improve Your Software Architecture with Ports and Adapters](https://spin.atomicobject.com/2013/02/23/ports-adapters-software-architecture/) by Tony Baker
+
+[Ports & Adapters Architecture](https://herbertograca.com/2017/09/14/ports-adapters-architecture/) by Herberto Graça
+
+[Ports-And-Adapters / Hexagonal Architecture](http://www.dossier-andreas.net/software_architecture/ports_and_adapters.html)
```
Comment (aRn0D) [inline TESTING.md]: you should add
* https://matthiasnoback.nl/2018/03/mocking-the-filesystem-and-randomness/
* https://matthiasnoback.nl/2018/02/mocking-at-architectural-boundaries-persistence-and-time/
* http://www.natpryce.com/articles/000714.html?utm_content=bufferf3fe8&utm_medium=social&utm_source=twitter.com&utm_campaign=buffer

Comment (jjanvier) [reply TESTING.md]: done in a new section

File: TESTING.md [hunk 13]
```diff
@@ -0,0 +1,237 @@
...
+
+- backend: test the generation of the completeness from a product
+- frontend: test the dropdown component
+
+### Acceptance - test a business use case or ensure a business rule
+
+Characteristics:
+
+- it tests several classes at the same time
+- it uses the business language, which means _Gherkin_ must be used
```
Comment (damien-carcel) [inline TESTING.md]: <3

File: TESTING.md [hunk 14]
```diff
@@ -0,0 +1,237 @@
...
+
+- backend/frontend: test the completeness has been calculated when the user filled in values in a product
+
+### Integration - test the integration of a brick with the outside world
+
+Characteristics:
+
+- it has no double (it tests the real classes)
+- it may test several classes at the same time
+- it tests only services that perform [I/O calls](#faq)
```
Comment (damien-carcel) [inline TESTING.md]: Maybe add that it basically tests stuff that we replace in acceptance tests.

File: TESTING.md [hunk 15]
```diff
@@ -0,0 +1,237 @@
...
+### Unit - test a unit of code
+
+Most of the time, a unit of code is a class.
+
+Characteristics:
+
+- it tests only one class at a time
+- it performs no [I/O call](#faq)
+- it doesn't require any setup (no fixtures for instance)
+- it doubles only what we own
```
Comment (rouliane) [inline TESTING.md]: I don't understand this sentence (probably because of my bad english level...). What does it mean ?

File: TESTING.md [hunk 16]
```diff
@@ -0,0 +1,237 @@
...
+
+Examples:
+
+- backend/frontend: test the completeness has been calculated when the user filled in values in a product
+
+### Integration - test the integration of a brick with the outside world
+
+Characteristics:
+
+- it has no double (it tests the real classes)
```
Comment (jjanvier) [inline TESTING.md]: double?
what about testing API calls with in memory repos
it still seems to be an integration test

File: TESTING.md [hunk 17]
```diff
@@ -0,0 +1,237 @@
...
+- frontend: test the dropdown component
+
+### Acceptance - test a business use case or ensure a business rule
+
+Characteristics:
+
+- it tests several classes at the same time
+- it uses the business language, which means _Gherkin_ must be used
+- it describes a business use case or it ensures a business rule (it's not about UI, CLI or UX, neither about a text we should see)
+- it doubles only what we own
```
[the remark this answers is not in the document]
Comment (ahocquard) [reply TESTING.md]: double is the generic term for mock, stub and fake. It is more appropriate to use it (we are not forced to use fake)

File: TESTING.md [hunk 18]
```diff
@@ -0,0 +1,237 @@
...
+- it treats the application as a black box
+- it has no double (it tests the real application)
+- it can require a complex setup (like a browser and Selenium for instance)
+- it tests nominal use cases
+- it tests critical use cases
+
+Tools:
+
+- Behat with Mink and Selenium
+
```
Comment (jjanvier) [inline TESTING.md]: missing phpunit

**Pull request closed without merging**